### PR TITLE
PROD-408 Routes for notification and bot user

### PR DESCRIPTION
**Pull request by neel-ds**
- Description
The PR adds two routes for notification handling and bot user handling. It includes schema changes required to maintain subscriber property. 

- What are the steps to test that this code is working?
Test the bot with /resubscribe and /unsubscribe commands

- Screen shots or recordings for UI changes
Screenshot 2024-10-30 at 2 36 53 AM

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -26,6 +27,7 @@ model User {
   fungibleAssetTransactionLog FungibleAssetTransactionLog[]
   revealNfts                  RevealNft[]
   tutorialCompletedAt         DateTime?
+  isSubscriber                Boolean                       @default(false)
```
Comment (marvinmarnold) [inline prisma/schema.prisma:30]: This variable name is not specific enough. Can you rename to `isBotSubscriber` or similar.

File: app/api/notification/route.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
+/*
+  THIS ROUTE CONTAINS TWO METHODS GET AND POST: 
+  ➤ GET METHOD: FETCHES SUBSCRIBED USERS
```
Comment (marvinmarnold) [inline app/api/notification/route.ts:3]: Let's split this up into two files. Paths should be more descriptive, eg `app/api/users/getSubscribedUsers/route.ts`

File: app/api/notification/route.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  ➤ POST METHOD: SETS A USER AS SUBSCRIBER
+*/
+import prisma from "@/app/services/prisma";
+import { headers } from "next/headers";
+
+export async function GET() {
```
Comment (marvinmarnold) [inline app/api/notification/route.ts:9]: All API methods should have tests

File: app/api/notification/route.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    return Response.json({ users });
+  } catch {
+    return new Response(`Error getting subscribed users`, {
+      status: 500,
+    });
+  }
+}
+
+export async function POST(request: Request) {
```
Comment (marvinmarnold) [inline app/api/notification/route.ts:38]: Unnecessary? Can't the user#GET route below be used for this?

**Review by marvinmarnold — CHANGES_REQUESTED** [review]

File: __tests__/api/user/createUserByTelegramId/route.test.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
+import { POST } from "@/app/api/user/createUserByTelegramId/route";
+import prisma from "@/app/services/prisma";
+import * as nodeCrypto from "crypto";
+import * as jwt from "jsonwebtoken";
+
+// Constants
+const VALID_API_KEY = process.env.BOT_API_KEY;
+const BOT_TOKEN = process.env.BOT_TOKEN ?? "";
+const BASE_URL = process.env.VERCEL_PROJECT_PRODUCTION_URL
```
Comment (marvinmarnold) [inline __tests__/api/user/createUserByTelegramId/route.test.ts:9]: You shouldn't need to define a remote endpoint. And should definitely not be using a prod endpoint. See: https://github.com/gator-labs/chomp/blob/7c80d01eebe9610b39ddd0f73f9e6f12b56d8861/__tests__/api/mixpanel/route.test.ts#L92

File: __tests__/api/user/createUserByTelegramId/route.test.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      expect(data.profile.wallets).toEqual([]);
+
+      // Store ID for cleanup
+      createdUserId = data.profile.id;
+
+      // Verify user exists in database
+      const dbUser = await prisma.user.findUnique({
+        where: { id: createdUserId },
+      });
+      expect(dbUser).toBeDefined();
```
Comment (marvinmarnold) [inline __tests__/api/user/createUserByTelegramId/route.test.ts:78]: Would be nice to make some additional assertions about the fields that we expect in dbUser.

File: __tests__/api/user/createUserByTelegramId/route.test.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+      // Assert
+      expect(response.status).toBe(500);
+      expect(errorMessage).toBe("Error creating user");
+    });
+  });
+});
+
+// Helper function to encode the Telegram payload
+export async function encodeTelegramPayload(TELEGRAM_USER_PAYLOAD: any) {
```
Comment (marvinmarnold) [inline __tests__/api/user/createUserByTelegramId/route.test.ts:127]: Would be nice to put this in a test utility file

File: __tests__/api/user/getUserByTelegramId/route.test.ts [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+describe("GET /api/user/getUserByTelegramId", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+    mockApiKey = VALID_API_KEY;
+  });
+
+  describe("when all parameters are valid", () => {
+    it("should return user profile with 200 status", async () => {
+      // Arrange
```
Comment (marvinmarnold) [inline __tests__/api/user/getUserByTelegramId/route.test.ts:58]: Nice, good to see the test split up into the three sections, for clarity.

File: __tests__/api/user/setUserSubscription/route.test.ts [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+const VALID_API_KEY = process.env.BOT_API_KEY;
+const TELEGRAM_USER_PAYLOAD = {
+  id: "705689",
+  first_name: "John",
+  last_name: "",
+  username: "john_doe",
+  photo_url: "",
+  auth_date: String(Math.floor(new Date().getTime())),
+};
+const BASE_URL = process.env.VERCEL_PROJECT_PRODUCTION_URL
```
Comment (marvinmarnold) [inline __tests__/api/user/setUserSubscription/route.test.ts:17]: Same issue that this shouldn't be necessary.

File: app/api/user/createUserByTelegramId/route.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+  const data = await request.json();
+  const { telegramAuthToken } = data;
+
+  if (!telegramAuthToken) {
+    return Response.json("Telegram payload is required", {
+      status: 400,
+    });
+  }
+
+  const randomUUID = uuidv4();
```
Comment (marvinmarnold) [inline app/api/user/createUserByTelegramId/route.ts:51]: This doesn't seem right to me. With our new direction, even TG users will be signed in before they use the app. We should never need to create a random UUID. It should always be available from Dynamic.

**Review by marvinmarnold — CHANGES_REQUESTED** [review]

File: __tests__/api/user/createUserByTelegramId/route.test.ts (hunk 5 above)
Comment (neel-ds) [reply __tests__/api/user/createUserByTelegramId/route.test.ts]: Good idea!

File: __tests__/api/user/createUserByTelegramId/route.test.ts (hunk 6 above)
Comment (neel-ds) [reply __tests__/api/user/createUserByTelegramId/route.test.ts:77]: Added telegramId assertion

File: app/api/user/createUserByTelegramId/route.ts (hunk 10 above)
Comment (neel-ds) [reply app/api/user/createUserByTelegramId/route.ts:51]: I went through the Dynamic APIs and am unsure if we can create a user based on TG credentials.

File: __tests__/api/user/getUserByTelegramId/route.test.ts [hunk 11]
```diff
@@ -0,0 +1,164 @@
+import verifyTelegramAuthToken from "@/app/actions/bot";
+import { GET } from "@/app/api/user/getUserByTelegramId/route";
+import prisma from "@/app/services/prisma";
+
+// Constants
+const VALID_API_KEY = process.env.BOT_API_KEY;
+const BASE_URL = process.env.VERCEL_PROJECT_PRODUCTION_URL
```
Comment (marvinmarnold) [inline __tests__/api/user/getUserByTelegramId/route.test.ts:7]: Why is this necessary?

File: prisma/migrations/20241029134128_add_tg_bot_fields/migration.sql [hunk 12]
```diff
@@ -0,0 +1,3 @@
+-- AlterTable
+ALTER TABLE "User" ADD COLUMN     "isBotSubscriber" BOOLEAN NOT NULL DEFAULT false,
+ADD COLUMN     "telegramId" BIGINT;
```
Comment (marvinmarnold) [inline prisma/migrations/20241029134128_add_tg_bot_fields/migration.sql:3]: Should be distinct

File: app/api/user/createUserByTelegramId/route.ts (hunk 10 above)
Comment (marvinmarnold) [reply app/api/user/createUserByTelegramId/route.ts:51]: @neel-ds will update auto login code to always update randomly generated user ID with a Dynamic ID.

File: app/actions/jwt.ts [hunk 13]
```diff
@@ -47,21 +47,69 @@ export const setJwt = async (token: string, nextPath?: string | null) => {
...
+      });
+    } else {
+      // If no existing user was found or the IDs match, create/update the user record
+      user = await prisma.user.upsert({
+        where: {
+          id: payload.sub,
+        },
+        create: {
+          id: payload.sub,
+          profileSrc: getRandomAvatarPath(),
```
Comment (marvinmarnold) [inline app/actions/jwt.ts:86]: This will result in us updating the profile image all the time.

File: app/actions/jwt.ts [hunk 14]
```diff
@@ -47,21 +47,69 @@ export const setJwt = async (token: string, nextPath?: string | null) => {
...
-    },
-    update: {},
-    include: {
-      wallets: true,
-      emails: true,
-    },
-  });
+  let user;
+
+  if (telegramUsername) {
```
Comment (marvinmarnold) [inline app/actions/jwt.ts:52]: I'm not sure this will keep working if a user changes their username

**Review by marvinmarnold — CHANGES_REQUESTED** [review]

File: app/actions/jwt.ts (hunk 13 above)
Comment (neel-ds) [reply app/actions/jwt.ts:86]: This is the existing implementation which I added in else statement. I think `upsert` didn't change the profile before as far as I know

**Comment by neel-ds** [comment]
I updated the logic for storing users based on Telegram ID instead of Telegram username.

File: app/actions/jwt.ts [hunk 15]
```diff
@@ -47,21 +51,71 @@ export const setJwt = async (token: string, nextPath?: string | null) => {
...
+    }
+  } else {
+    user = await prisma.user.upsert({
+      where: {
+        id: payload.sub,
+      },
+      create: {
+        id: payload.sub,
+        profileSrc: getRandomAvatarPath(),
+        telegramUsername: telegramUsername,
```
Comment (marvinmarnold) [inline app/actions/jwt.ts:110]: This will always be null here. Might as well remove, for clarity that nothing is being set.

**Review by marvinmarnold — CHANGES_REQUESTED** [review]

File: app/actions/jwt.ts [hunk 16]
```diff
@@ -47,21 +51,82 @@ export const setJwt = async (token: string, nextPath?: string | null) => {
...
+            id: payload.sub,
+            telegramUsername: telegramUsername,
+          },
+          include: {
+            wallets: true,
+            emails: true,
+          },
+        });
+      } else {
+        // If the user's Dynamic ID matches the JWT payload, then update their Telegram username incase user may have changed it
```
Comment (marvinmarnold) [inline app/actions/jwt.ts:85]: Why is it even necessary to update here? I would think that's only necessary if the username has changed.

**Comment by neel-ds** [comment]
I optimized the conditional statements to handle the user data store (create or update). No additional calls will be made in any conditions if not necessary.

File: app/actions/jwt.ts [hunk 17]
```diff
@@ -47,21 +51,75 @@ export const setJwt = async (token: string, nextPath?: string | null) => {
...
+        include: {
+          wallets: true,
+          emails: true,
+        },
+      });
+    }
+
+    // Update existing user if needed
+    const isTelegramUserDataChanged =
+      existingTelegramUser.id !== payload.sub ||
```
Comment (marvinmarnold) [inline app/actions/jwt.ts:95]: In what case would the tgId change? I don't think tgId should ever change.

**Review by marvinmarnold — COMMENTED** [review]
A ew more small questions.

File: app/api/user/setUserSubscription/route.ts [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+  const data = await request.json();
+  const { telegramAuthToken, isBotSubscriber } = data;
+
+  try {
+    const telegramUserData = (await verifyTelegramAuthToken(
+      telegramAuthToken,
+    )) as TelegramAuthDataProps;
+
+    const user = await prisma.user.findFirst({
```
[the remark this answers is not in the document]
Comment (neel-ds) [reply app/api/user/setUserSubscription/route.ts]: Good catch

File: app/api/user/setUserSubscription/route.ts [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+      where: { telegramId: telegramUserData.id },
+      select: {
+        id: true,
+      },
+    });
+
+    await prisma.user.update({
+      where: {
+        telegramUsername: telegramUserData.username,
```
[the remark this answers is not in the document]
Comment (neel-ds) [reply app/api/user/setUserSubscription/route.ts]: Done

File: app/actions/jwt.ts (hunk 17 above)
Comment (neel-ds) [reply app/actions/jwt.ts:95]: Telegram Id will always be the same.

File: app/actions/jwt.ts [hunk 20]
```diff
@@ -47,21 +51,75 @@ export const setJwt = async (token: string, nextPath?: string | null) => {
...
+        update: {},
+        include: {
+          wallets: true,
+          emails: true,
+        },
+      });
+    }
+
+    // Check for existing Telegram user
+    const existingTelegramUser = await prisma.user.findFirst({
```
Comment (yash431garg) [inline app/actions/jwt.ts:72]: I think you can use upsert here as well and don't need to use separately insert and update.

File: app/actions/jwt.ts [hunk 21]
```diff
@@ -47,21 +51,75 @@ export const setJwt = async (token: string, nextPath?: string | null) => {
...
+        },
+        include: {
+          wallets: true,
+          emails: true,
+        },
+      });
+    }
+
+    // If no updates needed, fetch current user
+    return prisma.user.findUnique({
```
Comment (yash431garg) [inline app/actions/jwt.ts:113]: Is this necessary?

**Review by yash431garg — COMMENTED** [review]
Can you check these comments? I think it will be fine after that only getUserByTelegram test will need some updates.

**Review by yash431garg — APPROVED** [review]

**Review by marvinmarnold — APPROVED** [review]

**Pull request merged**